### ISSUE TITLE
Jetpack Compose Coil gets stuck in Loading state in a LazyColumn that uses a placeholder

### ISSUE THREAD

**Issue by jakepurple13** [coil-kt/coil#1799]
**Describe the bug**
When going to a screen that has a LazyColumn, the images load. The problem is when going back a screen, then returning to the LazyColumn screen. I'm making use of [Accompanist's placeholder library](https://google.github.io/accompanist/placeholder/) as well. For some reason, coil is getting stuck in a loading state.

The code that enables the placeholder:
```kotlin
val painter = rememberAsyncImagePainter(
    model = imageUrl,
    placeholder = painterResource(R.drawable.ic_launcher_background),
    onSuccess = {
        println("JJJ Success!")
    },
    onLoading = {
        println("JJJ Loading!")
    },
    onError = {
        println("JJJ Error!")
        it.result.throwable.printStackTrace()
    }
)

Modifier.placeholder(
    visible = painter.state !is AsyncImagePainter.State.Success,
    color = PlaceholderDefaults.shimmerHighlightColor(),
    highlight = PlaceholderHighlight.shimmer()
)
```

**To Reproduce**
This project makes it easy to reproduce:
https://github.com/jakepurple13/CoilWeirdness
When running, press `Go` then press on an image or back and it will return to the `Go` screen. Then press `Go` again and stuck loading state happens.

**Logs/Screenshots**
https://github.com/coil-kt/coil/assets/2065574/5bb84723-659f-4c0b-b820-4b08b9e09589
![Screen Shot 2023-06-30 at 09 45 37](https://github.com/coil-kt/coil/assets/2065574/0ded091b-9bc2-470b-9ac1-92624ddff95e)

**Version**
Coil version: 2.4.0

**Comment by colinrtwhite** [comment]
Thanks for the repro project! Will take a look unless someone beats me to it.

**Comment by jakepurple13** [comment]
I took a look into this, since I had some free time, and I think I know what's causing the issue, kind of.
After MANY logs, I found that:
```kotlin
eventListener.resolveSizeStart(request).also { println("before size") }
val size = request.sizeResolver.size() //TODO: This is blocking it
eventListener.resolveSizeEnd(request, size).also { println("after size") }
```
the middle guy is blocking it.

If I change the `requestOf` function to have a set size:
```kotlin
@Composable
@ReadOnlyComposable
internal fun requestOf(model: Any?): ImageRequest {
    if (model is ImageRequest) {
        return model
    } else {
        return ImageRequest.Builder(LocalContext.current).size(400).data(model).build() //<-- here!!!
    }
}
```
everything works as intended. But that is just a work around.
I plan on continuing to see what could possibly be blocking it and why.

**Comment by jakepurple13** [comment]
I figured out what the problem is!
```kotlin
private val drawSize = MutableStateFlow(Size.Zero)
```
Since that guy starts with `Size.Zero`, when calling this:
```kotlin
size { drawSize.mapNotNull { it.toSizeOrNull() }.first() }
```
`toSizeOrNull()`'s check is returning `null`!
```kotlin
private fun Size.toSizeOrNull() = when {
    isUnspecified -> CoilSize.ORIGINAL
    isPositive -> CoilSize(
        width = if (width.isFinite()) Dimension(width.roundToInt()) else Dimension.Undefined,
        height = if (height.isFinite()) Dimension(height.roundToInt()) else Dimension.Undefined
    )
    else -> null
}
```
Since it ends up going to the else condition.
When I set the default size to:
```kotlin
private val drawSize = MutableStateFlow(Size.Unspecified)
```
it works without issue!

**Comment by colinrtwhite** [comment]
Thanks for taking a look! The `drawSize` call looks to be working correctly as we want it to return `null` until it's given a valid size. I'd check out [here](https://github.com/coil-kt/coil/blob/main/coil-compose-base/src/main/java/coil/compose/AsyncImage.kt#L207) as that callback should be invoked by Compose to provide the correct constraints which are used to set `drawSize`.

**Comment by jakepurple13** [comment]
I have found that if I do:
```kotlin
val painter = rememberAsyncImagePainter(
            model = ImageRequest.Builder(LocalContext.current)
                .data(imageUrl)
                .size(Size.ORIGINAL)
                .build()
)
```
It works completely fine.

**Comment by colinrtwhite** [comment]
@jakepurple13 Setting `size(Size.ORIGINAL)` skips waiting for Compose to call [this callback](https://github.com/coil-kt/coil/blob/main/coil-compose-base/src/main/java/coil/compose/AsyncImage.kt#L207).

**Comment by jakepurple13** [comment]
Yeah, the problem is that I'm not using AsyncImage. I'm using a normal image and it only happens when a placeholder is used. Unless I'm misunderstanding something which could be the case here.

**Comment by colinrtwhite** [comment]
This is fixed in Coil 3.0 as `rememberAsyncImagePainter` no longer waits for `onDraw` to be called before launching the request.
